engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/GeneralAnnotations.scala

```diff
@@ -0,0 +1,89 @@
+package org.enso.compiler.pass.resolve
+
+import org.enso.compiler.context.{InlineContext, ModuleContext}
+import org.enso.compiler.core.IR
+import org.enso.compiler.core.IR.Module.Scope.Definition
+import org.enso.compiler.core.ir.MetadataStorage._
+import org.enso.compiler.exception.CompilerError
+import org.enso.compiler.pass.IRPass
+
+/** A pass responsible for the discovery of [[IR.Name.GeneralAnnotation]]
+  * annotations, and for associating them with the corresponding construct.
+  *
+  * Compilation pipeline of annotations:
+  * - [[ModuleAnnotations]] pass ignores general annotations and leaves them in
+  * the tree so that the consequent passes are able to process the annotation
+  * expression.
+  * - [[org.enso.compiler.pass.desugar.ComplexType]] pass associates general
+  * annotations with the type constructor definitions.
+  * - [[GeneralAnnotations]] pass associates general annotations that are left
+  * in the tree with the appropriate definitions.
+  */
+case object GeneralAnnotations extends IRPass {
+  override type Metadata = ModuleAnnotations.Annotations
+  override type Config   = IRPass.Configuration.Default
+  override val precursorPasses: Seq[IRPass]   = Seq()
+  override val invalidatedPasses: Seq[IRPass] = Seq()
+
+  /** Resolves annotations.
+    *
+    * @param ir the Enso IR to process
+    * @param moduleContext a context object that contains the information needed
+    * to process a module
+    * @return `ir`, possibly having made transformations or annotations to that
+    * IR.
+    */
+  override def runModule(
+    ir: IR.Module,
+    moduleContext: ModuleContext
+  ): IR.Module = {
+    var lastAnnotations: Seq[IR.Name.GeneralAnnotation] = Seq()
+    val newBindings = ir.bindings.map {
```

Comment:
minor nit: this feels like it could be rewritten using `fold` to avoid this `lastAnnotations` reference, cleaning it later etc.

Reply:
Here we're filtering `ir.binding` as well as mutating the `lastAnnotations` reference. While it can be done with `foldLeft`, I think the imperative way is a bit cleaner